### PR TITLE
[DNM / EXPERIMENTAL] Re-specialize row arrangement key closure

### PR DESCRIPTION
**Pull request by vmarcos**
This PR reintroduces the closure that was special-purposed for forming row-row arrangement key-values. By doing so, it aims at establishing if the alternative based on arrangement specialization traits introduces regret by higher abstraction.

### Motivation

Investigate https://github.com/MaterializeInc/materialize/issues/22885.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by vmarcos** [comment]
Relevant experimental results in [Slack](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1699316245860339?thread_ts=1699004491.151929&cid=C02PPB50ZHS).

**Comment by vmarcos** [comment]
TL;DR: #22887 properly fixed the additional memory pressure caused by the arrangement specialization framework (no difference wrt. using the old closure demonstrable by this PR). A second part of the observed memory spike in repros, as reported in #22918, can be attributed to changes that made `persist_source` more efficient at pushing data out (ref #22549). This accentuates the spike caused in downstream operators that need to buffer (as in the case of arranging to hydrate an index), even if it may reduce memory requirements of `persist_source` itself.
